File: tests/agent_workloads/adapters/git_diff_review.mjs
import { execFile } from "node:child_process";

if (process.argv.length !== 3) {
  throw new Error("usage: git_diff_review INPUT_ROOT");
}

function gitDiff(root) {
  return new Promise((resolve, reject) => {
    execFile(
      "git",
      [
        "-C",
        root,
        "-c",
        "core.quotePath=false",
        "diff",
        "--no-index",
        "--no-renames",
        "--name-status",
        "--",
        "before",
        "after",
      ],
      {
        encoding: "utf8",
        timeout: 5000,
        killSignal: "SIGKILL",
        maxBuffer: 1024 * 1024,
      },
      (error, stdout, stderr) => {
        if (error?.code === 1) {
          resolve(stdout);
        } else {
          reject(new Error(`git diff exit ${error?.code ?? 0}: ${stderr.trim()}`));
        }
      },
    );
  });
}

const kinds = { A: "added", D: "deleted", M: "modified" };
const counts = { A: 0, D: 0, M: 0 };
const rows = [];
for (const line of (await gitDiff(process.argv[2])).split(/\r?\n/)) {
  if (!line) continue;
  const fields = line.split("\t");
  if (fields.length !== 2 || !(fields[0] in kinds)) {
    throw new Error(`bad git name-status row: ${line}`);
  }
  const [status, rawPath] = fields;
  let path;
  if (rawPath.startsWith("before/")) {
    path = rawPath.slice("before/".length);
  } else if (rawPath.startsWith("after/")) {
    path = rawPath.slice("after/".length);
  } else {
    throw new Error(`git path escaped roots: ${rawPath}`);
  }
  counts[status] += 1;
  rows.push(`${path}|${kinds[status]}`);
}

for (const row of rows.sort()) {
  console.log(row);
}
console.log(
  `summary|added=${counts.A}|modified=${counts.M}|deleted=${counts.D}`,
);

File: tests/agent_workloads/adapters/git_diff_review.py
import subprocess
import sys
from pathlib import Path


if len(sys.argv) != 2:
    raise SystemExit("usage: git_diff_review INPUT_ROOT")

root = Path(sys.argv[1])
completed = subprocess.run(
    [
        "git",
        "-C",
        str(root),
        "-c",
        "core.quotePath=false",
        "diff",
        "--no-index",
        "--no-renames",
        "--name-status",
        "--",
        "before",
        "after",
    ],
    capture_output=True,
    text=True,
    timeout=5,
    check=False,
)
if completed.returncode != 1:
    raise SystemExit(f"git diff exit {completed.returncode}: {completed.stderr.strip()}")

counts = {"A": 0, "D": 0, "M": 0}
kinds = {"A": "added", "D": "deleted", "M": "modified"}
rows = []
for line in completed.stdout.splitlines():
    fields = line.split("\t")
    if len(fields) != 2 or fields[0] not in kinds:
        raise SystemExit(f"bad git name-status row: {line}")
    status, path = fields
    if path.startswith("before/"):
        path = path[len("before/") :]
    elif path.startswith("after/"):
        path = path[len("after/") :]
    else:
        raise SystemExit(f"git path escaped roots: {path}")
    counts[status] += 1
    rows.append(f"{path}|{kinds[status]}")

print(*sorted(rows), sep="\n")
print(f"summary|added={counts['A']}|modified={counts['M']}|deleted={counts['D']}")
